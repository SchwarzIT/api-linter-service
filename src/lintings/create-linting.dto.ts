import { ApiProperty } from '@nestjs/swagger';
import { IsBase64, IsNotEmpty } from 'class-validator';
import { ApiType } from '../helpers/enum.api-type';

export class CreateLintingDto {
  @ApiProperty({
    description: 'The API type to check.',
    example: 'product_api',
    enum: Object.keys(ApiType),
  })
  @IsNotEmpty()
  apiType: keyof typeof ApiType;

  @ApiProperty({
    description: 'The base64 representation of your API spec.',
    example:
      'eyJvcGVuYXBpIjoiMy4wLjAiLCJwYXRocyI6eyIvZGlnaXRhbC10d2luL2FwaS92MS9wcm9kdWN0cyI6eyJwb3N0Ijp7Im9wZXJhdGlvbklkIjoiUHJvZHVjdHNDb250cm9sbGVyX2NyZWF0ZSIsInN1bW1hcnkiOiIiLCJkZXNjcmlwdGlvbiI6IkNyZWF0ZSBhIG5ldyBwcm9kdWN0LiIsInBhcmFtZXRlcnMiOltdLCJyZXF1ZXN0Qm9keSI6eyJyZXF1aXJlZCI6dHJ1ZSwiY29udGVudCI6eyJhcHBsaWNhdGlvbi9qc29uIjp7InNjaGVtYSI6eyIkcmVmIjoiIy9jb21wb25lbnRzL3NjaGVtYXMvQ3JlYXRlUHJvZHVjdER0byJ9fX19LCJyZXNwb25zZXMiOnsiMjAxIjp7ImRlc2NyaXB0aW9uIjoiQ3JlYXRlZCBuZXcgcHJvZHVjdCBhbmQgcmVzcG9uZHMgdGhlIG5ld2x5IGNyZWF0ZWQgcHJvZHVjdC4iLCJjb250ZW50Ijp7ImFwcGxpY2F0aW9uL2pzb24iOnsic2NoZW1hIjp7IiRyZWYiOiIjL2NvbXBvbmVudHMvc2NoZW1hcy9DcmVhdGVkUHJvZHVjdER0byJ9fX19LCI0MDEiOnsiZGVzY3JpcHRpb24iOiJBUEkgY2FsbCB3YXMgbm90IHN1Y2Nlc3NmdWxseSB2YWxpZGF0ZWQgYWdhaW5zdCBBenVyZUFELiBHaXZlbiBBdXRob3JpemF0aW9uIFRva2VuIGlzIG5vdCB2YWxpZC4ifSwiNDA5Ijp7ImRlc2NyaXB0aW9uIjoiUHJvZHVjdCB3aXRoIGdpdmVuIGVhbiBhbHJlYWR5IGV4aXN0cyJ9fSwidGFncyI6WyJwcm9kdWN0cyJdLCJzZWN1cml0eSI6W3siYmVhcmVyIjpbXX1dfSwiZ2V0Ijp7Im9wZXJhdGlvbklkIjoiUHJvZHVjdHNDb250cm9sbGVyX2ZpbmRBbGwiLCJzdW1tYXJ5IjoiIiwiZGVzY3JpcHRpb24iOiJHZXQgYWxsIHByb2R1Y3RzLiIsInBhcmFtZXRlcnMiOltdLCJyZXNwb25zZXMiOnsiMjAwIjp7ImRlc2NyaXB0aW9uIjoiUmVzcG9uZHMgYWxsIHByb2R1Y3RzIGFzIGEgbGlzdCBvZiBwcm9kdWN0cy4iLCJjb250ZW50Ijp7ImFwcGxpY2F0aW9uL2pzb24iOnsic2NoZW1hIjp7InR5cGUiOiJhcnJheSIsIml0ZW1zIjp7IiRyZWYiOiIjL2NvbXBvbmVudHMvc2NoZW1hcy9DcmVhdGVkUHJvZHVjdER0byJ9fX19fSwiNDAxIjp7ImRlc2NyaXB0aW9uIjoiQVBJIGNhbGwgd2FzIG5vdCBzdWNjZXNzZnVsbHkgdmFsaWRhdGVkIGFnYWluc3QgQXp1cmVBRC4gR2l2ZW4gQXV0aG9yaXphdGlvbiBUb2tlbiBpcyBub3QgdmFsaWQuIn0sIjQwNCI6eyJkZXNjcmlwdGlvbiI6Ik5vIHByb2R1Y3RzIGZvdW5kLiJ9fSwidGFncyI6WyJwcm9kdWN0cyJdLCJzZWN1cml0eSI6W3siYmVhcmVyIjpbXX1dfX0sIi9kaWdpdGFsLXR3aW4vYXBpL3YxL3Byb2R1Y3RzL3tlYW59Ijp7ImdldCI6eyJvcGVyYXRpb25JZCI6IlByb2R1Y3RzQ29udHJvbGxlcl9maW5kT25lIiwic3VtbWFyeSI6IiIsImRlc2NyaXB0aW9uIjoiR2V0IHByb2R1Y3QgYnkgZWFuLiIsInBhcmFtZXRlcnMiOlt7Im5hbWUiOiJlYW4iLCJyZXF1aXJlZCI6dHJ1ZSwiaW4iOiJwYXRoIiwic2NoZW1hIjp7InR5cGUiOiJzdHJpbmcifX1dLCJyZXNwb25zZXMiOnsiMjAwIjp7ImRlc2NyaXB0aW9uIjoiUmVzcG9uZHMgb25lIHByb2R1Y3Qgd2l0aCBnaXZlbiBcImVhblwiIiwiY29udGVudCI6eyJhcHBsaWNhdGlvbi9qc29uIjp7InNjaGVtYSI6eyIkcmVmIjoiIy9jb21wb25lbnRzL3NjaGVtYXMvQ3JlYXRlZFByb2R1Y3REdG8ifX19fSwiNDAxIjp7ImRlc2NyaXB0aW9uIjoiQVBJIGNhbGwgd2FzIG5vdCBzdWNjZXNzZnVsbHkgdmFsaWRhdGVkIGFnYWluc3QgQXp1cmVBRC4gR2l2ZW4gQXV0aG9yaXphdGlvbiBUb2tlbiBpcyBub3QgdmFsaWQuIn0sIjQwNCI6eyJkZXNjcmlwdGlvbiI6IlByb2R1Y3Qgd2l0aCBnaXZlbiBlYW4gd2FzIG5vdCBmb3VuZC4ifX0sInRhZ3MiOlsicHJvZHVjdHMiXSwic2VjdXJpdHkiOlt7ImJlYXJlciI6W119XX0sInB1dCI6eyJvcGVyYXRpb25JZCI6IlByb2R1Y3RzQ29udHJvbGxlcl91cGRhdGUiLCJzdW1tYXJ5IjoiIiwiZGVzY3JpcHRpb24iOiJVcGRhdGUgYW4gZXhpc3RpbmcgcHJvZHVjdC4iLCJwYXJhbWV0ZXJzIjpbeyJuYW1lIjoiZWFuIiwicmVxdWlyZWQiOnRydWUsImluIjoicGF0aCIsInNjaGVtYSI6eyJ0eXBlIjoic3RyaW5nIn19XSwicmVxdWVzdEJvZHkiOnsicmVxdWlyZWQiOnRydWUsImNvbnRlbnQiOnsiYXBwbGljYXRpb24vanNvbiI6eyJzY2hlbWEiOnsiJHJlZiI6IiMvY29tcG9uZW50cy9zY2hlbWFzL1VwZGF0ZVByb2R1Y3REdG8ifX19fSwicmVzcG9uc2VzIjp7IjIwMCI6eyJkZXNjcmlwdGlvbiI6IlVwZGF0ZWQgZ2l2ZW4gcHJvZHVjdCBhbmQgcmVzcG9uZHMgdGhlIHVwZGF0ZWQgcHJvZHVjdC4iLCJjb250ZW50Ijp7ImFwcGxpY2F0aW9uL2pzb24iOnsic2NoZW1hIjp7IiRyZWYiOiIjL2NvbXBvbmVudHMvc2NoZW1hcy9DcmVhdGVkUHJvZHVjdER0byJ9fX19LCI0MDEiOnsiZGVzY3JpcHRpb24iOiJBUEkgY2FsbCB3YXMgbm90IHN1Y2Nlc3NmdWxseSB2YWxpZGF0ZWQgYWdhaW5zdCBBenVyZUFELiBHaXZlbiBBdXRob3JpemF0aW9uIFRva2VuIGlzIG5vdCB2YWxpZC4ifSwiNDA0Ijp7ImRlc2NyaXB0aW9uIjoiUHJvZHVjdCB3aXRoIGdpdmVuIGVhbiB3YXMgbm90IGZvdW5kIGFuZCBjb3VsZCB0aGVyZWZvcmUgbm90IGJlIHVwZGF0ZWQuIn19LCJ0YWdzIjpbInByb2R1Y3RzIl0sInNlY3VyaXR5IjpbeyJiZWFyZXIiOltdfV19LCJkZWxldGUiOnsib3BlcmF0aW9uSWQiOiJQcm9kdWN0c0NvbnRyb2xsZXJfcmVtb3ZlIiwic3VtbWFyeSI6IiIsImRlc2NyaXB0aW9uIjoiRGVsZXRlIGFuIGV4aXN0aW5nIHByb2R1Y3QuIiwicGFyYW1ldGVycyI6W3sibmFtZSI6ImVhbiIsInJlcXVpcmVkIjp0cnVlLCJpbiI6InBhdGgiLCJzY2hlbWEiOnsidHlwZSI6InN0cmluZyJ9fV0sInJlc3BvbnNlcyI6eyIyMDAiOnsiZGVzY3JpcHRpb24iOiJEZWxldGVkIGdpdmVuIHByb2R1Y3QuIn0sIjQwMSI6eyJkZXNjcmlwdGlvbiI6IkFQSSBjYWxsIHdhcyBub3Qgc3VjY2Vzc2Z1bGx5IHZhbGlkYXRlZCBhZ2FpbnN0IEF6dXJlQUQuIEdpdmVuIEF1dGhvcml6YXRpb24gVG9rZW4gaXMgbm90IHZhbGlkLiJ9LCI0MDQiOnsiZGVzY3JpcHRpb24iOiJQcm9kdWN0IHdpdGggZ2l2ZW4gZWFuIHdhcyBub3QgZm91bmQgYW5kIGNvdWxkIHRoZXJlZm9yZSBub3QgYmUgZGVsZXRlZC4ifX0sInRhZ3MiOlsicHJvZHVjdHMiXSwic2VjdXJpdHkiOlt7ImJlYXJlciI6W119XX19fSwiaW5mbyI6eyJ0aXRsZSI6IkRpZ2l0YWwgVHdpbiBBUEkiLCJkZXNjcmlwdGlvbiI6IlRoZSBEaWdpdGFsIFR3aW4gQVBJIGRlc2NyaXB0aW9uIiwidmVyc2lvbiI6IjEuMCIsImNvbnRhY3QiOnt9fSwidGFncyI6W10sInNlcnZlcnMiOlt7InVybCI6Imh0dHA6Ly9sb2NhbGhvc3QifSx7InVybCI6Imh0dHA6Ly9xcyJ9LHsidXJsIjoiaHR0cDovL3Byb2QifV0sImNvbXBvbmVudHMiOnsic2VjdXJpdHlTY2hlbWVzIjp7ImJlYXJlciI6eyJzY2hlbWUiOiJiZWFyZXIiLCJiZWFyZXJGb3JtYXQiOiJKV1QiLCJ0eXBlIjoiaHR0cCJ9fSwic2NoZW1hcyI6eyJDcmVhdGVQcm9kdWN0RHRvIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7ImVhbiI6eyJ0eXBlIjoic3RyaW5nIiwiZGVzY3JpcHRpb24iOiJUaGUgdW5pcXVlIHByb2R1Y3QgZWFuLiIsImV4YW1wbGUiOiIxMjEyMzMyNCJ9LCJkZXNjcmlwdGlvbiI6eyJ0eXBlIjoic3RyaW5nIiwiZGVzY3JpcHRpb24iOiJQcm9kdWN0IGRlc2NyaXB0aW9uLiIsImV4YW1wbGUiOiJUaGlzIGlzIGEgdmVyeSBuaWNlIHByb2R1Y3QgZGVzY3JpcHRpb24uIn0sIm1hbnVmYWN0dXJlciI6eyJ0eXBlIjoic3RyaW5nIiwiZGVzY3JpcHRpb24iOiJQcm9kdWN0IG1hbnVmYWN0dXJlci4iLCJleGFtcGxlIjoiS2luZGVyIn0sImJyYW5kIjp7InR5cGUiOiJzdHJpbmciLCJkZXNjcmlwdGlvbiI6IlByb2R1Y3QgYnJhbmQuIiwiZXhhbXBsZSI6IktpbmRlciBTY2hva29sYWRlIn19LCJyZXF1aXJlZCI6WyJlYW4iLCJkZXNjcmlwdGlvbiIsIm1hbnVmYWN0dXJlciJdfSwiQ3JlYXRlZFByb2R1Y3REdG8iOnsidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsiZWFuIjp7InR5cGUiOiJzdHJpbmciLCJkZXNjcmlwdGlvbiI6IlRoZSB1bmlxdWUgcHJvZHVjdCBlYW4uIiwiZXhhbXBsZSI6IjEyMTIzMzI0In0sImRlc2NyaXB0aW9uIjp7InR5cGUiOiJzdHJpbmciLCJkZXNjcmlwdGlvbiI6IlByb2R1Y3QgZGVzY3JpcHRpb24uIiwiZXhhbXBsZSI6IlRoaXMgaXMgYSB2ZXJ5IG5pY2UgcHJvZHVjdCBkZXNjcmlwdGlvbi4ifSwibWFudWZhY3R1cmVyIjp7InR5cGUiOiJzdHJpbmciLCJkZXNjcmlwdGlvbiI6IlByb2R1Y3QgbWFudWZhY3R1cmVyLiIsImV4YW1wbGUiOiJLaW5kZXIifSwiYnJhbmQiOnsidHlwZSI6InN0cmluZyIsImRlc2NyaXB0aW9uIjoiUHJvZHVjdCBicmFuZC4iLCJleGFtcGxlIjoiS2luZGVyIFNjaG9rb2xhZGUifSwiaWQiOnsidHlwZSI6InN0cmluZyIsImRlc2NyaXB0aW9uIjoiUHJvZHVjdCB1dWlkLiIsImV4YW1wbGUiOiJlZmQ0MzJhYy1kNGJhLTQ5ZDMtOTI1Yy05MDI0ZTM3ZTFhZjYifX0sInJlcXVpcmVkIjpbImVhbiIsImRlc2NyaXB0aW9uIiwibWFudWZhY3R1cmVyIiwiYnJhbmQiLCJpZCJdfSwiVXBkYXRlUHJvZHVjdER0byI6eyJ0eXBlIjoib2JqZWN0IiwicHJvcGVydGllcyI6eyJlYW4iOnsidHlwZSI6InN0cmluZyIsImRlc2NyaXB0aW9uIjoiVGhlIHVuaXF1ZSBwcm9kdWN0IGVhbi4iLCJleGFtcGxlIjoiMTIxMjMzMjQifSwiZGVzY3JpcHRpb24iOnsidHlwZSI6InN0cmluZyIsImRlc2NyaXB0aW9uIjoiUHJvZHVjdCBkZXNjcmlwdGlvbi4iLCJleGFtcGxlIjoiVGhpcyBpcyBhIHZlcnkgbmljZSBwcm9kdWN0IGRlc2NyaXB0aW9uLiJ9LCJtYW51ZmFjdHVyZXIiOnsidHlwZSI6InN0cmluZyIsImRlc2NyaXB0aW9uIjoiUHJvZHVjdCBtYW51ZmFjdHVyZXIuIiwiZXhhbXBsZSI6IktpbmRlciJ9LCJicmFuZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVzY3JpcHRpb24iOiJQcm9kdWN0IGJyYW5kLiIsImV4YW1wbGUiOiJLaW5kZXIgU2Nob2tvbGFkZSJ9fX19fSwiZXh0ZXJuYWxEb2NzIjp7ImRlc2NyaXB0aW9uIjoiQ29tcGFueSBBUEkgYmVzdCBwcmFjdGljZXMiLCJ1cmwiOiJodHRwczovL3ByZXNzLXJlbGVhc2UtZGVtby5wcm9kLnNpdC5zeXMub2RqLmNsb3VkL2FyY2hpdGVjdHVyZS1iZXN0LXByYWN0aWNlcy9hcGlzLyJ9fQ==',
  })
  @IsBase64()
  @IsNotEmpty()
  apiSpecAsBase64: string;
}
