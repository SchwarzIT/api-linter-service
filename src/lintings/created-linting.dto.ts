import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { SpectralResult } from './spectralResult.dto';

export class CreatedLintingDto {
  @ApiProperty({
    description:
      'Information if your OpenApi Spec does comply with company rules.',
    example: 'Given API Spec DOES comply with company API rules.',
  })
  description: string;

  @ApiProperty({
    description: 'Link to Company API Best Practices',
    example:
      'https://press-release-demo.prod.sit.sys.odj.cloud/architecture-best-practices/apis/',
  })
  linkApiRules: string;

  @ApiPropertyOptional({
    description: 'Highest severity level during linting.',
    example: 'Warn',
    enum: ['Error', 'Warn', 'Info', 'Hint'],
  })
  highestSeverityLevel?: string;

  @ApiProperty({
    description:
      'Indicates if given OpenAPI Spec complies to company API best practices.',
    example: true,
  })
  isValidSpec: boolean;

  @ApiProperty({
    description: 'Example API linting result',
    type: () => SpectralResult,
    isArray: true,
    example: [],
  })
  lintingResults: SpectralResult[];
}
