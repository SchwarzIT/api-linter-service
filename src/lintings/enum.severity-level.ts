export enum SeverityLevel {
  Error,
  Warn,
  Info,
  Hint,
}
