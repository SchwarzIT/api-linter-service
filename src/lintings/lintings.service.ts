import { Injectable } from '@nestjs/common';
import {
  Spectral,
  Ruleset,
  ISpectralDiagnostic,
} from '@stoplight/spectral-core';
import { ApiType } from '../helpers/enum.api-type';
import { CreateLintingDto } from './create-linting.dto';
import { CreatedLintingDto } from './created-linting.dto';
import { SeverityLevel } from './enum.severity-level';

@Injectable()
export class LintingsService {
  private spectral: Spectral;

  async validateApiSpec(
    createLintingDto: CreateLintingDto,
  ): Promise<CreatedLintingDto> {
    const apiType: ApiType = ApiType[createLintingDto.apiType];
    const rules: Ruleset = await import(`../spectral-${apiType}`);
    this.spectral = new Spectral();
    this.spectral.setRuleset(rules);
    const swaggerFile = this.getSwaggerFileAsString(
      createLintingDto.apiSpecAsBase64,
    );
    const lintingResult = await this.getLintingResult(swaggerFile);
    const hasNoLintingErrors = this.hasNoLintingErrors(lintingResult);
    return this.calculateResponse(hasNoLintingErrors, lintingResult);
  }

  private hasNoLintingErrors(lintingResult: ISpectralDiagnostic[]): boolean {
    let errors = 0;
    lintingResult.forEach((partialResult) => {
      if (
        partialResult.severity === 0 ||
        partialResult.code === 'unrecognized-format'
      ) {
        errors++;
      }
    });
    return errors === 0;
  }

  private getSwaggerFileAsString(swaggerFileAsBase64): string {
    return Buffer.from(swaggerFileAsBase64, 'base64').toString('ascii');
  }

  private calculateResponse(
    hasNoLintingErrors: boolean,
    lintingResult: ISpectralDiagnostic[],
  ): CreatedLintingDto {
    if (!hasNoLintingErrors) {
      const failedResponse: CreatedLintingDto = {
        isValidSpec: false,
        description: 'Given API Spec DOES NOT comply with company API rules.',
        linkApiRules:
          'https://onedirection.schwarz/architecture-best-practices/apis/',
        highestSeverityLevel: SeverityLevel[0],
        lintingResults: lintingResult,
      };
      return failedResponse;
    }
    const response: CreatedLintingDto = {
      isValidSpec: true,
      description: 'Given API Spec DOES comply with company API rules.',
      linkApiRules:
        'https://onedirection.schwarz/architecture-best-practices/apis/',
      lintingResults: lintingResult,
    };
    return response;
  }

  private async getLintingResult(swaggerFile): Promise<ISpectralDiagnostic[]> {
    return this.spectral.run(swaggerFile);
  }
}
