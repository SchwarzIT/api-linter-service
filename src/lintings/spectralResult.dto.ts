import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { ISpectralDiagnostic } from '@stoplight/spectral-core';
import {
  JsonPath,
  IRange,
  DiagnosticSeverity,
  IDiagnosticRelatedInformation,
  IPosition,
} from '@stoplight/types';

export class SpectralResult implements ISpectralDiagnostic {
  @ApiProperty({
    description: 'The path of the error as JsonPath',
    isArray: true,
    type: 'string',
    example: ['paths', '/route', 'get'],
  })
  path: JsonPath;

  @ApiProperty({
    description:
      'The code of the rule, can be used to identify the rule that caused the issue',
    example: 'path-description-is-mandatory',
    oneOf: [{ type: 'string' }, { type: 'number' }],
  })
  code: string | number;

  @ApiProperty({
    type: () => SpectralResultRange,
    description: 'The range of the error inside the linted document',
    example: {
      start: { line: 14, character: 21 },
      end: { line: 84, character: 52 },
    } as SpectralResultRange,
  })
  range: IRange;

  @ApiPropertyOptional({
    description: 'The message associated with the issue',
    example:
      'Every route of an API should have a description; property: /route.description is missing',
  })
  message: string;

  @ApiProperty({
    description:
      'The severity of the issue. 0 = "Error", 1 = "Warning", 2 = "Information", 3 = "Hint"',
    enum: [0, 1, 2, 3],
  })
  severity: DiagnosticSeverity;

  @ApiPropertyOptional({
    description:
      'A human-readable string describing the source of this diagnostic, e.g. "typescript" or "super lint".',
  })
  source?: string;

  @ApiPropertyOptional({
    description: 'Additional metadata about the diagnostic.',
  })
  tags?: string[];

  relatedInformation?: IDiagnosticRelatedInformation[];
}

export class SpectralResultRange implements IRange {
  @ApiProperty({
    type: () => SpectralResultPosition,
    description:
      'The starting position of the issue inside the linted document',
  })
  start: IPosition;

  @ApiProperty({
    type: () => SpectralResultPosition,
    description: 'The ending position of the issue inside the linted document',
  })
  end: IPosition;
}

export class SpectralResultPosition implements IPosition {
  @ApiProperty({
    description: 'Line position in a document (zero-based)',
  })
  line: number;

  @ApiProperty({
    description:
      'Character offset on a line in a document (zero-based). Assuming that the line is represented as a string, the `character` value represents the gap between the `character` and `character + 1`. If the character value is greater than the line length it defaults back to the line length.',
  })
  character: number;
}
