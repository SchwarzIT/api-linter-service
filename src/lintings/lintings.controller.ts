import { BadRequestException, Body, Controller, Post } from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiBasicAuth,
  ApiCreatedResponse,
  ApiOperation,
  ApiTags,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger';
import { LintingsService } from './lintings.service';
import { CreateLintingDto } from './create-linting.dto';
import { CreatedLintingDto } from './created-linting.dto';
import { ApiType } from '../helpers/enum.api-type';

@Controller('/api-linting/api/v1/lintings')
export class LintingsController {
  constructor(private readonly lintingsService: LintingsService) {}

  @ApiOperation({
    description: 'Create a new API linting.',
  })
  @ApiCreatedResponse({
    description: 'Created new API linting.',
    type: CreatedLintingDto,
  })
  @ApiBadRequestResponse({
    description: 'Provided request body is missing mandatory properties.',
  })
  @ApiUnauthorizedResponse({
    description: 'API call was not authenticated.',
  })
  @ApiTags('lintings')
  @ApiBasicAuth()
  @Post()
  createLinting(
    @Body() createLintingDto: CreateLintingDto,
  ): Promise<CreatedLintingDto> {
    if (!ApiType[createLintingDto.apiType]) {
      throw new BadRequestException('Property "apiType" not provided in body.');
    }
    return this.lintingsService.validateApiSpec(createLintingDto);
  }
}
