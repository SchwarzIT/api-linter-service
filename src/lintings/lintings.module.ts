import { Module } from '@nestjs/common';
import { LintingsController } from './lintings.controller';
import { LintingsService } from './lintings.service';

@Module({
  imports: [],
  controllers: [LintingsController],
  providers: [LintingsService],
})
export class LintingsModule {}
