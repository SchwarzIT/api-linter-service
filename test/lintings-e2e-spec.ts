import { Test, TestingModule } from '@nestjs/testing';
import { HttpStatus, INestApplication } from '@nestjs/common';
import * as supertest from 'supertest';
import { CreateLintingDto } from '../src/lintings/create-linting.dto';
import { LintingsModule } from '../src/lintings/lintings.module';
import { CreatedLintingDto } from 'src/lintings/created-linting.dto';
import validationPipe from 'src/helpers/validation.pipe';
import { mockUsername, mockPassword, mockPasswordHash } from './fixtures/auth';
import { apiSpecAsBase64 } from './fixtures/apiSpec';

describe('[Feature] Lintings - /api-linting/api/v1/lintings', () => {
  let app: INestApplication;
  const requestBody: CreateLintingDto = {
    apiType: 'product_api',
    apiSpecAsBase64,
  };

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [LintingsModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(validationPipe());
    await app.init();
  });

  describe('POST', () => {
    it('successfully creates a new linting', async () => {
      return supertest(app.getHttpServer())
        .post('/api-linting/api/v1/lintings')
        .auth(mockUsername, mockPassword)
        .send(requestBody)
        .expect(HttpStatus.CREATED)
        .then(({ body }: { body: CreatedLintingDto }) => {
          expect(body.description).toBe(
            'Given API Spec DOES comply with company API rules.',
          );
          expect(body.linkApiRules).toBe(
            'https://onedirection.schwarz/architecture-best-practices/apis/',
          );
          expect(body).toMatchSnapshot('lintingResultSwaggerFile');
        });
    });

    it('failes with a bad request response if the apiType is missing', () => {
      return supertest(app.getHttpServer())
        .post('/api-linting/api/v1/lintings')
        .auth(mockUsername, mockPassword)
        .send({ apiSpecAsBase64: 'spec' })
        .expect(HttpStatus.BAD_REQUEST);
    });

    it('failes with a bad request response if the apiSpec is missing', () => {
      return supertest(app.getHttpServer())
        .post('/api-linting/api/v1/lintings')
        .auth(mockUsername, mockPassword)
        .send({ apiType: 'product_api' })
        .expect(HttpStatus.BAD_REQUEST);
    });

    it('failes when the request is lacking authorization', () => {
      return supertest(app.getHttpServer())
        .post('/api-linting/api/v1/lintings')
        .send(requestBody)
        .expect(HttpStatus.UNAUTHORIZED);
    });
  });

  afterAll(async () => {
    await app.close();
  });
});
