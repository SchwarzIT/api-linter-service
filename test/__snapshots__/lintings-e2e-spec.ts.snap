// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Feature] Lintings - /api-linting/api/v1/lintings POST successfully creates a new linting: lintingResultSwaggerFile 1`] = `
Object {
  "description": "Given API Spec DOES comply with company API rules.",
  "highestSeverityLevel": "Warn",
  "isValidSpec": true,
  "linkApiRules": "https://onedirection.schwarz/architecture-best-practices/apis/",
  "lintingResults": Array [
    Object {
      "code": "info-description",
      "message": "OpenAPI object info \`description\` must be present and at least 100 chars long.",
      "path": Array [
        "info",
        "description",
      ],
      "range": Object {
        "end": Object {
          "character": 3244,
          "line": 0,
        },
        "start": Object {
          "character": 3210,
          "line": 0,
        },
      },
      "severity": 1,
    },
    Object {
      "code": "contact-information",
      "message": "Every API must have a contact containing name, email and a url; property contact.name is missing",
      "path": Array [
        "info",
        "contact",
      ],
      "range": Object {
        "end": Object {
          "character": 3273,
          "line": 0,
        },
        "start": Object {
          "character": 3271,
          "line": 0,
        },
      },
      "severity": 1,
    },
    Object {
      "code": "contact-information",
      "message": "Every API must have a contact containing name, email and a url; property contact.email is missing",
      "path": Array [
        "info",
        "contact",
      ],
      "range": Object {
        "end": Object {
          "character": 3273,
          "line": 0,
        },
        "start": Object {
          "character": 3271,
          "line": 0,
        },
      },
      "severity": 1,
    },
    Object {
      "code": "contact-information",
      "message": "Every API must have a contact containing name, email and a url; property contact.url is missing",
      "path": Array [
        "info",
        "contact",
      ],
      "range": Object {
        "end": Object {
          "character": 3273,
          "line": 0,
        },
        "start": Object {
          "character": 3271,
          "line": 0,
        },
      },
      "severity": 1,
    },
    Object {
      "code": "must-define-example-schema",
      "message": "Every DTO must define at least one example; DTO is lacking an example #/components/schemas/CreateProductDto",
      "path": Array [
        "components",
        "schemas",
        "CreateProductDto",
      ],
      "range": Object {
        "end": Object {
          "character": 3956,
          "line": 0,
        },
        "start": Object {
          "character": 3494,
          "line": 0,
        },
      },
      "severity": 1,
    },
    Object {
      "code": "must-define-example-schema",
      "message": "Every DTO must define at least one example; DTO is lacking an example #/components/schemas/CreatedProductDto",
      "path": Array [
        "components",
        "schemas",
        "CreatedProductDto",
      ],
      "range": Object {
        "end": Object {
          "character": 4556,
          "line": 0,
        },
        "start": Object {
          "character": 3979,
          "line": 0,
        },
      },
      "severity": 1,
    },
    Object {
      "code": "must-define-example-schema",
      "message": "Every DTO must define at least one example; DTO is lacking an example #/components/schemas/UpdateProductDto",
      "path": Array [
        "components",
        "schemas",
        "UpdateProductDto",
      ],
      "range": Object {
        "end": Object {
          "character": 4991,
          "line": 0,
        },
        "start": Object {
          "character": 4578,
          "line": 0,
        },
      },
      "severity": 1,
    },
  ],
}
`;
